pm/recipient.go

```diff
@@ -134,20 +165,80 @@ func (r *recipient) RedeemWinningTickets(sessionIDs []string) error {
 }
 
 // TicketParams returns the recipient's currently accepted ticket parameters
-func (r *recipient) TicketParams(sender ethcommon.Address) *TicketParams {
+func (r *recipient) TicketParams(sender ethcommon.Address) (*TicketParams, error) {
 	randBytes := RandBytes(32)
 
 	seed := new(big.Int).SetBytes(randBytes)
 	recipientRand := r.rand(seed, sender)
 	recipientRandHash := crypto.Keccak256Hash(ethcommon.LeftPadBytes(recipientRand.Bytes(), uint256Size))
 
+	faceValue, err := r.faceValue(sender)
+	if err != nil {
+		return nil, err
+	}
+
 	return &TicketParams{
 		Recipient:         r.addr,
-		FaceValue:         r.faceValue,
-		WinProb:           r.winProb,
+		FaceValue:         faceValue,
+		WinProb:           r.winProb(faceValue),
 		RecipientRandHash: recipientRandHash,
 		Seed:              seed,
+	}, nil
+}
+
+func (r *recipient) faceValue(sender ethcommon.Address) (*big.Int, error) {
+	gasPrice := r.gpm.GasPrice()
+	// txCost = redeemGas * gasPrice
+	txCost := new(big.Int).Mul(big.NewInt(int64(r.cfg.RedeemGas)), gasPrice)
+
+	// faceValue = txCost * txCostMultiplier
+	faceValue := new(big.Int).Mul(txCost, big.NewInt(int64(r.cfg.TxCostMultiplier)))
+
+	// TODO: Consider setting faceValue to some value higher than
+	// EV in this case where the default faceValue < the desired EV.
+	// At the moment, for simplicity we just adjust faceValue to the
+	// desired EV in this case (which would result in winProb = 100%).
+	// In practice, EV should be smaller than the default faceValue
```

Comment:
Could you help me understand why EV should be smaller then the default `faceValue`? Perhaps point me to some documentation that might be helpful?

Reply:
Essentially `EV = faceValue * winProb`

So a ticket with a faceValue of `500 wei` and a winProb of `10%` would have an expected payout per ticket of `50 wei` 

Now let's say our EV is 500 wei and our defaultFaceValue is 250 wei we'd get 
`500 = 250 * winProb`
which results in a winProbability of 200%. 

That's of course not possible which is why if `defaultFaceValue < EV` we adjust the faceValue to the EV. Resulting in a winning probability of 100%, which is still not optimal but from talking to @yondonfu he considers that an edge case currently.

I'm probably missing quite a bit of essential context here that @yondonfu could provide better than me I think. But I think it essentially boils down to:

> _No expectation of long lived relationships with some minimum value transacted:_ An orchestrator is not affected by broadcasters that send a few non-winning tickets and then disconnect because it knows that it will eventually receive a sufficient number of tickets from the numerous broadcasters it works for such that it will be paid fairly. Furthermore, since orchestrators set the face value of tickets, they can always ensure that it will be profitable to redeem received winning tickets

From the [livepeer blog on PM](https://medium.com/livepeer-blog/streamflow-probabilistic-micropayments-f3a647672462)

Reply:
Again, makes a lot of sense! Thanks @kyriediculous. Basically, the only time winning probability should be 100% is when an O doesn't understand the concept of PM? No one would rationally choose to have every ticket be a winning ticket, which would almost always mean a higher transaction cost than ticket value

Reply:
Still trying to wrap my mind around it sometimes as well. But yes that would mean an Orchestrator doesn't really understand the concept. I think most Orchestrators would be using the default value anyway and changing defaults would be for advanced users only. 

There wouldn't directly be a higher tx cost than ticket value, but high winning probabilities could add to network congestion leads to having to use a higher gasPrice in order to keep the same confirmation time for transactions. 

A higher winning probability would also require the orchestrator to redeem his tickets more frequently to increase the chance of actually getting paid in full based on the outstanding float from a broadcaster to an orchestrator. 

Eg. 
Round 1: 10 O's
B's Reserve: 5 eth (0.5 eth per Orch allocated) 
B's Deposit: 1 eth

- In Round 1 O receives a winning ticket (100% winProb) for 1.5 eth but he waits to redeem it until the next round because the gasPrice is currently high

Now let's say in the next round there are 20 O's , then the allocation for each O from B would drop to 0.25eth (for simplicity sake, no one has redeemed a ticket from B in the meantime, so deposit remains at 1 eth) 

B's deposit + allocation from B to O = 1.25 eth now, which is not enough to cover the full payment of O's winning ticket of 1.5eth sent in the previous round. 

From the [PM spec](https://github.com/livepeer/wiki/blob/master/spec/streamflow/pm.md#reserve): 

> An orchestrator can accept or reject work from a broadcaster based upon the broadcaster's reserve and the value of the allocation from the reserve committed to the orchestrator. For example, since the value of the allocation affects the maximum face value that can be used for tickets, lower reserves would require lower ticket face values and higher winning probabilities - if the winning probability is too high such that an orchestrator would need to redeem winning tickets too frequently (potentially contributing to network congestion), then an orchestrator might reject work from the broadaster.

Reply:
In the context of payment accounting, an orchestrator's EV is also the minimum credit requirement imposed on a broadcaster (as described [here](https://github.com/livepeer/go-livepeer/issues/866)) - given an orchestrator EV of X, an orchestrator will only transcode segments if the broadcaster's current credit balance is at least X. 

The minimum credit requirement feature allows an orchestrator to control how much value it is guaranteed when it transcodes segments. A higher EV = a higher minimum credit requirement which is "safer" for an orchestrator (in case transcoding a segment is costly). However, at the same time, a higher EV is more "risky" for a broadcaster because if a broadcaster has to send X value to an orchestrator up front, the broadcaster stands to lose X if the orchestrator does not respond. So, in practice, the EV should be high enough to cover some reasonable amount of transcoding work for the orchestrator and low enough so the broadcaster doesn't put a lot of value at risk.

If the EV is higher than the default faceValue (which by default is 100x the current transaction cost of ticket redemption), then that would represent a pretty high minimum credit requirement:

Ex. 

```
redeemGas = 100000
gasPrice = 5 GWEI
txCost = 0.0005 ETH
faceValue = 100 * txCost = .05 ETH -> $12.5 @250 ETH/USD
```

So, if the EV is equal to the default faceValue here, then the minimum credit requirement would be $12.5 and the broadcaster would be putting $12.5 at risk up front which is pretty unreasonable and shouldn't happen in practice as long as the broadcaster software is configured properly to reject orchestrators that set an absurdly high EV. In practice, the EV should be on the magnitude of cents or fractions of cents. Consider the following purely illustrative example using per segment pricing (we're planning on using per pixel pricing instead), AWS Elastic Transcoder pricing results in something like $.003 per 2 second segment. So, a reasonable EV/minimum credit requirement covering 3 segments would be 3 x $.003 = $.009 -> .9 cents. I think a reasonable EV will generally be closer to this range of values than it would be to the `faceValue`. Higher EV values that approach the default faceValue would be more likely to be rejected by broadcasters. Thus, the case where EV > default faceValue feels like an edge case that we can handle in a simplistic manner for this PR and consider more optimal logic later (i.e. if the EV > default faceValue, set a new default faceValue that is some multiple of the EV).